smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/commit/KafkaRedisCheckpointCommit.java

```diff
@@ -0,0 +1,112 @@
+
+    private RedisAPI redisAPI;
+
+    public KafkaRedisCheckpointCommit(KafkaConnectorIncomingConfiguration config,
+            Vertx vertx,
+            KafkaConsumer<?, ?> consumer,
+            BiConsumer<Throwable, Boolean> reportFailure,
+            int defaultTimeout,
+            Redis redis) {
+        super(vertx, config, consumer, reportFailure, defaultTimeout);
+        this.redis = redis;
+    }
+
+    private <T> Uni<T> runWithRedis(Function<RedisAPI, Uni<T>> action) {
+        return Uni.createFrom().item(this.redisAPI)
+                .onItem().ifNull().switchTo(this.redis.connect().map(RedisAPI::api).onItem().invoke(api -> this.redisAPI = api))
+                .flatMap(action::apply);
+    }
+
+    @ApplicationScoped
+    @Identifier("checkpoint-redis")
+    public static class Factory implements KafkaCommitHandler.Factory {
+
+        @Override
+        public KafkaCommitHandler create(KafkaConnectorIncomingConfiguration config, Vertx vertx,
+                KafkaConsumer<?, ?> consumer, BiConsumer<Throwable, Boolean> reportFailure) {
+            int defaultTimeout = config.config()
+                    .getOptionalValue(ConsumerConfig.DEFAULT_API_TIMEOUT_MS_CONFIG, Integer.class)
+                    .orElse(60000);
+            String redisConnectionString = config.config().getValue("redis-connection-string", String.class);
+            Redis redis = Redis.createClient(vertx, new RedisOptions()
+                    .setMaxPoolSize(30)
+                    .setMaxWaitingHandlers(1024)
+                    .addConnectionString(redisConnectionString));
+
+            return new KafkaRedisCheckpointCommit(config, vertx, consumer, reportFailure, defaultTimeout, redis);
+        }
+
+    }
+
+    @Override
+    public void terminate(boolean graceful) {
+        super.terminate(graceful);
+        if (redisAPI != null) {
+            redisAPI.close();
+        }
+        redis.close();
+    }
+
+    @Override
+    protected Uni<ProcessingState<?>> fetchProcessingState(TopicPartition partition) {
+        return runWithRedis(api -> api.get(getKey(partition))
+                .onFailure().invoke(t -> log.errorf(t, "Error fetching processing state %s", partition))
+                .onItem().invoke(r -> log.debugf("Fetched state for partition %s : %s", partition, r))
+                .map(r -> Optional.ofNullable(r)
+                        .map(Response::toBuffer)
+                        .map(this::deserializeState)
+                        .orElse(null)));
+    }
+
+    private String getKey(TopicPartition partition) {
+        return partition.topic() + "-" + partition.partition();
+    }
+
+    private <T> ProcessingState<T> deserializeState(Buffer b) {
+        return Json.decodeValue(b.getDelegate(), ProcessingState.class);
+    }
+
+    @Override
+    protected Uni<Void> persistProcessingState(TopicPartition partition, ProcessingState<?> state) {
+        // TODO perform watch, get, offset check, multi, set, discard/exec
```

Comment:
you are going to love that, you need to implement a check-and-set construct, as inside a transaction, commands do not produce results.

Reply:
I'd tested an implementation but couldn't get it working. I am thinking now that my test was broken. I've pushed an implementation. Let's see what the CI says.

Reply:
Note that this implementation is on tests only. We can decide later on if it'll be exposed to the public API from this module.